### PR TITLE
[rllib] Asynchronous PPO with IMPALA architecture

### PR DESCRIPTION
**Pull request by michaelzhiluo**
Closes https://github.com/ray-project/ray/issues/1015

### CODE REVIEW

File: python/ray/rllib/agents/ppo/vtrace.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
+# Copyright 2018 Google LLC
```
[the remark this answers is not in the document]
Comment (michaelzhiluo) [reply python/ray/rllib/agents/ppo/vtrace.py]: Done

File: python/ray/rllib/tuned_examples/pong-impala.yaml [hunk 2]
```diff
@@ -1,13 +1,13 @@
-# This can reach 18-19 reward within 10 minutes on a Tesla M60 GPU (e.g., G3 EC2 node):
-#   128 workers -> 8 minutes
```
[the remark this answers is not in the document]
Comment (michaelzhiluo) [reply python/ray/rllib/tuned_examples/pong-impala.yaml]: Its the same as old pong-impala.yaml

**Comment by ericl** [comment]
I think we should try to inherit most components besides the loss itself from impala, to make it more maintainable.

**Comment by michaelzhiluo** [comment]
Cleaned up PPO @ericl

**Comment by ericl** [comment]
Nice. Can you also merge that patch that adds an option to use the normal advantages? I'll work on cleaning this up more after that.

File: python/ray/rllib/agents/ppo/appo.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+import time
+
+from ray.rllib.agents.a3c.a3c_tf_policy_graph import A3CPolicyGraph
+from ray.rllib.agents.ppo.vtrace_policy_graph import VTracePolicyGraph
+from ray.rllib.agents.agent import Agent, with_common_config
+from ray.rllib.optimizers import AsyncSamplesOptimizer
+from ray.rllib.agents import impala
+
+OPTIMIZER_SHARED_CONFIGS = [
```
Comment (ericl) [inline python/ray/rllib/agents/ppo/appo.py]: Can just import the impala config here

File: python/ray/rllib/agents/ppo/appo.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    "max_sample_requests_in_flight_per_worker",
+    "broadcast_interval",
+    "num_sgd_iter",
+    "minibatch_buffer_size",
+]
+
+
+# yapf: disable
+# __sphinx_doc_begin__
+DEFAULT_CONFIG = with_common_config({
```
Comment (ericl) [inline python/ray/rllib/agents/ppo/appo.py]: Can you also have this extend the impala config (see apex.py for how to merge dicts).

File: python/ray/rllib/models/catalog.py [hunk 5]
```diff
@@ -226,6 +226,8 @@ def _get_model(input_dict, obs_space, num_outputs, options, state_in,
         obs_rank = len(input_dict["obs"].shape) - 1
 
         if obs_rank > 1:
+            if options["use_autoencoder"]:
```
Comment (ericl) [inline python/ray/rllib/models/catalog.py]: Remove this

**Pull request closed without merging**